packages/ethereum/contracts/ERC777/ERC777Snapshot.sol

```diff
@@ -14,171 +11,147 @@ import "@openzeppelin/contracts/token/ERC777/ERC777.sol";
+            // mint
+            updateValueAtNow(accountSnapshots[to], balanceOf(to).add(amount));
+            updateValueAtNow(totalSupplySnapshots, totalSupply().add(amount));
+        } else if (to == address(0)) {
+            // burn
+            updateValueAtNow(accountSnapshots[from], balanceOf(from).sub(amount));
+            updateValueAtNow(totalSupplySnapshots, totalSupply().sub(amount));
+        } else if (from != to) {
+            // transfer
+            updateValueAtNow(accountSnapshots[from], balanceOf(from).sub(amount));
+            updateValueAtNow(accountSnapshots[to], balanceOf(to).add(amount));
         }
     }
 
-    function _updateAccountSnapshot(address account) private {
-        _updateSnapshot(_accountBalanceSnapshots[account], balanceOf(account));
-    }
-
-    function _updateTotalSupplySnapshot() private {
-        _updateSnapshot(_totalSupplySnapshots, totalSupply());
-    }
+    /**
+     * @dev `_valueAt` retrieves the number of tokens at a given block number
+     * @param snapshots The history of values being queried
+     * @param _block The block number to retrieve the value at
+     * @return The number of tokens being queried
+     */
+    function _valueAt(
+        Snapshot[] storage snapshots,
+        uint _block
+    ) view internal returns (uint) {
+        if (snapshots.length == 0) return 0;
+
+        // Shortcut for the actual value
+        if (_block >= snapshots[snapshots.length - 1].fromBlock) {
+            return snapshots[snapshots.length - 1].value;
+        }
+        if (_block < snapshots[0].fromBlock) {
+            return 0;
+        }
 
-    function _updateSnapshot(Snapshots storage snapshots, uint256 currentValue) private {
-        uint256 currentId = _currentSnapshotId.current();
-        if (_lastSnapshotId(snapshots.ids) < currentId) {
-            snapshots.ids.push(currentId);
-            snapshots.values.push(currentValue);
+        // Binary search of the value in the array
+        uint min = 0;
+        uint max = snapshots.length - 1;
+        while (max > min) {
+            uint mid = (max + min + 1) / 2;
+            if (snapshots[mid].fromBlock <= _block) {
+                min = mid;
+            } else {
+                max = mid - 1;
+            }
         }
+        return snapshots[min].value;
     }
 
-    function _lastSnapshotId(uint256[] storage ids) private view returns (uint256) {
-        if (ids.length == 0) {
-            return 0;
+    /**
+     * @dev `updateValueAtNow` used to update the `balances` map and the
+     *  `totalSupplySnapshots`
+     * @param snapshots The history of data being updated
+     * @param _value The new number of tokens
+     */
+    function updateValueAtNow(Snapshot[] storage snapshots, uint256 _value) internal {
+        require(_value <= uint128(-1), "casting overflow");
+
+        if (
+            (snapshots.length == 0) ||
+            (snapshots[snapshots.length - 1].fromBlock < block.number)
```

Comment:
doesn't this need an `uint128` cast?

Reply:
I did some tests and I don't see any difference when we do comparison, checks out fine.